waiter/src/waiter/core.clj

```diff
@@ -285,6 +285,48 @@
       (catch Exception e
         (utils/exception->response e request)))))
 
+(defn add-grpc-headers-and-trailers
+  "Finds and attaches the equivalent grpc status codes for the provided http status code."
+  [{:keys [headers status trailers waiter/message] :as response}]
+  (if (utils/waiter-generated-response? response)
+    (if-let [grpc-status-data (cond
+                                (= status 400) ["3" "Bad Request"]
+                                (= status 401) ["16" "Unauthorized"]
+                                (= status 403) ["7" "Permission Denied"]
+                                (= status 429) ["14" "Too Many Requests"]
+                                (= status 500) ["13" "Internal Server Error"]
+                                (= status 502) ["14" "Bad Gateway"]
+                                (= status 503) ["14" "Service Unavailable"]
+                                (= status 504) ["4" "Gateway Timeout"])]
+      (let [[grpc-status standard-message] grpc-status-data
+            grpc-message (if (string? message) message standard-message)
+            trailers-ch (async/promise-chan)
+            new-headers (assoc headers
+                          "content-type" "application/grpc"
+                          "grpc-message" grpc-message
+                          "grpc-status" grpc-status)]
+        (async/go
```

Comment:
Do we expect to ever have trailers?  Do gRPC responses always require trailers?

Reply:
gRPC responses do not always require trailers, when both headers and trailers are present, it looks into the trailers for the status.

We need to send the trailers as a workaround to jetty limtation of closing the stream with an empty data frame when only headers are sent. The presence of a data frame confuses grpc and it reports a different error than the one sent by Waiter:

HTTP/2 traffic:
```
2019-07-26 01:46:04 DEBUG io.grpc.netty.NettyClientHandler - [id: 0x878392ea, L:/127.0.0.1:60283 - R:/127.0.0.1:9091] INBOUND HEADERS: streamId=3 headers=GrpcHttp2ResponseHeaders[:status: 504, x-cid: wgttguc150048199015780, grpc-status: 4, grpc-message: Request to service backend timed out, ...] padding=0 endStream=false
2019-07-26 01:46:04 DEBUG io.grpc.netty.NettyClientHandler - [id: 0x878392ea, L:/127.0.0.1:60283 - R:/127.0.0.1:9091] INBOUND DATA: streamId=3 padding=0 endStream=true length=0 bytes=
```

Status code:
```
{:code "INTERNAL", :description "Received unexpected EOS on DATA frame from server."}
```